lib/src/read-contents/read-buffer.js

```diff
@@ -1,23 +1,41 @@
 'use strict';
 
 var fs = require('graceful-fs');
+var iconv = require('iconv-lite');
 var removeBomBuffer = require('remove-bom-buffer');
 
+var getCodec = require('../../codecs');
+
 function bufferFile(file, optResolver, onRead) {
+  var encoding = optResolver.resolve('encoding', file);
+  var codec = getCodec(encoding);
+  if (encoding && !codec) {
+    return onRead(new Error('Unsupported encoding: ' + encoding));
+  }
+
   fs.readFile(file.path, onReadFile);
 
-  function onReadFile(readErr, data) {
+  function onReadFile(readErr, contents) {
     if (readErr) {
       return onRead(readErr);
     }
 
-    var removeBOM = optResolver.resolve('removeBOM', file);
-    if (removeBOM) {
-      file.contents = removeBomBuffer(data);
-    } else {
-      file.contents = data;
+    if (encoding) {
+      var removeBOM = codec.bomAware && optResolver.resolve('removeBOM', file);
+
+      if (codec.enc !== 'utf8') {
+        contents = iconv.decode(contents, encoding, { stripBOM: false });
+        removeBOM = removeBOM && contents[0] === '\ufeff';
+        contents = iconv.encode(contents, 'utf8');
```

Comment:
I hope I've addressed the translation question adequately, and I think it applies here as well.

As for the `fs.readFile` thing, it's similar to `fs.writeFile`: Node has a horrible API, in my humble opinion, in its efforts to "help us" :-)

https://nodejs.org/dist/latest-v8.x/docs/api/fs.html#fs_fs_readfile_path_options_callback
https://nodejs.org/dist/latest-v8.x/docs/api/fs.html#fs_fs_writefile_file_data_options_callback

If we omit the `encoding` argument, these methods will yield/expect (readFile/writeFile respectively) a Buffer containing just the raw unadulterated bytes. However if we provide an encoding argument, then suddenly these methods yield/expect Strings, which it decodes/encodes (readFile/writeFile) for us.

Long story short, better to not pass `encoding` to the `fs` methods and rely on iconv to do this for us.